### ISSUE TITLE
[REGRESSION] MoEs are obtaining higher loss than they should during training

### ISSUE THREAD

**Issue by kiddyboots216** [NVIDIA/Megatron-LM#894]
**Describe the regression**
In the forks of Megatron-LM used by gpt-neox and megatron-deepspeed, MoEs are obtaining lower loss than they are in Megatron-LM with the same configuration.

**To Reproduce**
Attached to this issue are config files to reproduce the exact MoE we are running. `megatron_125M_k1_e16_moe_3e-4_config.sh` is the MoE config for megatron-lm, `megatron_dense_125M_config.sh` is the dense config for megatron-lm, `gpt-neox_e16_k1_config.yaml` is the MoE config for gpt-neox. All models are gpt-style. 
[megatron_dense_125M_config.txt](https://github.com/user-attachments/files/16019016/megatron_dense_125M_config.txt)
[megatron_125M_k1_e16_moe_3e-4_config.txt](https://github.com/user-attachments/files/16019017/megatron_125M_k1_e16_moe_3e-4_config.txt)
[gpt-neox_e16_k1_config.txt](https://github.com/user-attachments/files/16019018/gpt-neox_e16_k1_config.txt)

**Previous performance**
After step `12000` in gpt-neox the MoE has training loss `2.452`.

**New performance**
After step `12000` in megatron-lm the MoE has training loss `2.649` which is the same as the dense model.

**Stack trace/logs**
The logs are attached to this issue.

**Environment (please complete the following information):**
 - Previous Megatron-LM commit ID: the version of Megatron-LM forked by gpt-neox / megatron-deepspeed
 - New Megatron-LM commit ID: the current commit ID; `e33c8f78a35765d5aa37475a144da60e8a2349d1`
 - Previous PyTorch version: same PyTorch version
 - New PyTorch version: same PyTorch version
 - Previous CUDA version: same CUDA version
 - New CUDA version: same CUDA version
 - Previous NCCL version: same NCCL version
 - New NCCL version: same NCCL version

**Proposed fix**
No proposed fix.

**Additional context**
Presumably a bug was introduced in the MoE training. However, I looked into the `gpt` code in mcore.models and was unable to find any potential causes.

**Comment by bentherien** [comment]
Here is some additional information:

The figure below shows validation loss curves for 125M MoE and dense models trained with [megatron_125M_k1_e16_moe_3e-4_config.txt](https://github.com/user-attachments/files/16019017/megatron_125M_k1_e16_moe_3e-4_config.txt) (maxLR \in {1e-4. 3e-4, 6e-4, 9e-4}) and [megatron_dense_125M_config.txt](https://github.com/user-attachments/files/16019016/megatron_dense_125M_config.txt), respectively. We observe that the MoE models underperform the dense models contrary to results form the literature. As mentioned above, this suggests that there is a bug in Megatron-LM

![e16-k1-moe-maxlr3e-4_vs_dense-maxlr3e-4](https://github.com/NVIDIA/Megatron-LM/assets/39810288/774a11df-534e-41c2-a939-375e6a8fd0e6)


Moreover, below is a plot directly comparing the training loss of dense and MoE models in Megatron and GPT-NeoX trained using GBS=768, SL=2048, E=16 (total exps), K=1 (active exps). All models are trained using the same dataset and the same linear warmup+consine annealing LRS (maxLR3e-4 to minLR3e-5). We observe that the GPT-NeoX implementation has results in line with the literature (e.g., [switch transformer](https://arxiv.org/abs/2101.03961) Figure1 right), while the Megatron implementation does not.

This suggests there is a bug in Megatron-LM @jaredcasper @duncanriach @jon-barker 

![megatron_and_neox_comparison](https://github.com/NVIDIA/Megatron-LM/assets/39810288/0c12ba2e-84c7-42a1-8521-3e122eabb6e9)

**Comment by kiddyboots216** [comment]
<img width="1405" alt="125M_exps8-val-loss" src="https://github.com/NVIDIA/Megatron-LM/assets/22651617/c57a4590-3c34-4a20-be6e-a63103b867a1">
Here is the validation loss plot for more MoE configs, again with varying LRs that all underperform the dense model.

**Comment by yqli2420** [comment]
> Here is some additional information:
> 
> The figure below shows validation loss curves for 125M MoE and dense models trained with [megatron_125M_k1_e16_moe_3e-4_config.txt](https://github.com/user-attachments/files/16019017/megatron_125M_k1_e16_moe_3e-4_config.txt) (maxLR \in {1e-4. 3e-4, 6e-4, 9e-4}) and [megatron_dense_125M_config.txt](https://github.com/user-attachments/files/16019016/megatron_dense_125M_config.txt), respectively. We observe that the MoE models underperform the dense models contrary to results form the literature. As mentioned above, this suggests that there is a bug in Megatron-LM
> 
> ![e16-k1-moe-maxlr3e-4_vs_dense-maxlr3e-4](https://private-user-images.githubusercontent.com/39810288/343891398-774a11df-534e-41c2-a939-375e6a8fd0e6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk1NTQ5MjIsIm5iZiI6MTcxOTU1NDYyMiwicGF0aCI6Ii8zOTgxMDI4OC8zNDM4OTEzOTgtNzc0YTExZGYtNTM0ZS00MWMyLWE5MzktMzc1ZTZhOGZkMGU2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI4VDA2MDM0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNmZWM1OTBkMDBmODVkZGFmNTBjMTM2NWQ4OTcwYmE0Y2RmNWU3NjlhYjc5ZTBmMmQyYmYyNTViZmIxMDc2MDMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.l4kN3epgb1z8fWMgGQ_yKp_6XaXawpjQwjdiPqaMYvU)

Personally, I think we should change the horizontal axis to FLOPs and then compare the loss.

**Comment by kiddyboots216** [comment]
> > Here is some additional information:
> 
> > 
> 
> > The figure below shows validation loss curves for 125M MoE and dense models trained with [megatron_125M_k1_e16_moe_3e-4_config.txt](https://github.com/user-attachments/files/16019017/megatron_125M_k1_e16_moe_3e-4_config.txt) (maxLR \in {1e-4. 3e-4, 6e-4, 9e-4}) and [megatron_dense_125M_config.txt](https://github.com/user-attachments/files/16019016/megatron_dense_125M_config.txt), respectively. We observe that the MoE models underperform the dense models contrary to results form the literature. As mentioned above, this suggests that there is a bug in Megatron-LM
> 
> > 
> 
> > ![e16-k1-moe-maxlr3e-4_vs_dense-maxlr3e-4](https://private-user-images.githubusercontent.com/39810288/343891398-774a11df-534e-41c2-a939-375e6a8fd0e6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk1NTQ5MjIsIm5iZiI6MTcxOTU1NDYyMiwicGF0aCI6Ii8zOTgxMDI4OC8zNDM4OTEzOTgtNzc0YTExZGYtNTM0ZS00MWMyLWE5MzktMzc1ZTZhOGZkMGU2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjI4VDA2MDM0MlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNmZWM1OTBkMDBmODVkZGFmNTBjMTM2NWQ4OTcwYmE0Y2RmNWU3NjlhYjc5ZTBmMmQyYmYyNTViZmIxMDc2MDMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.l4kN3epgb1z8fWMgGQ_yKp_6XaXawpjQwjdiPqaMYvU)
> 
> 
> 
> Personally, I think we should change the horizontal axis to FLOPs and then compare the loss.

These MoEs are all K=1, so they are already FLOPS-matched (in other words the plots would be the same if we changed the horizontal axis to FLOPS.)

**Comment by zainsarwar865** [comment]
Running the same config in Megatron-Deepspeed also results in the MoE models being better than their dense counterparts.

**Comment by zainsarwar865** [comment]
Running the same config with Megatron-DeepSpeed does result in the MoE outperforming the dense model. This was run with 8 experts, topk=1 and a 125M base model. 

<img width="1339" alt="Deepspeed_moe_dense_e-8_topk1" src="https://github.com/NVIDIA/Megatron-LM/assets/66789976/a4ee6c4d-7422-4ff7-bcd8-cd4bc1370567">

**Comment by yanring** [comment]
Thank you for reporting the issue! We will investigate it and get back to you soon.

**Comment by bentherien** [comment]
Thank you @yanring !

**Comment by yanring** [comment]
Hi @kiddyboots216 @bentherien ,

We have done some investigations and discovered that the issue specifically pertains to the Top-1 selection, and the root cause is the ordering of softmax and topk. In short, we should apply softmax before selecting the top-k if k equals 1, since performing softmax on [num_tokens, 1] would result in a gradient of 0. Below is our experiments and code changes:
<img width="1196" alt="image" src="https://github.com/NVIDIA/Megatron-LM/assets/12545681/628860f7-01c1-40c3-8446-409bec6755e5">
<img width="1044" alt="image" src="https://github.com/NVIDIA/Megatron-LM/assets/12545681/e5e8b026-a711-419c-9dc2-44c05876c2ed">

**Comment by zainsarwar865** [comment]
Thanks @yanring . I did observe that with top-k =2, the results were significantly better than what the literature suggests in terms of being better than top-k 1.

**Comment by yanring** [comment]
FYI: We have added the argument `--moe-router-pre-softmax` on [commit](https://github.com/NVIDIA/Megatron-LM/commit/5153efea0bc8e9f0cd6094d558d467e44622405b)

Thank you all!
